specs/experimental/gov-delegation.md

```diff
@@ -0,0 +1,440 @@
+defined as in the `GovernanceToken` and use the same types:
+
+```solidity
+// Addresses that had their delegation state migrated from the `GovernanceToken` to the `GovernanceDelegation`.
+mapping(address => bool) public migrated;
+
+// Voting power delegations of an account.
+mapping(address => Delegation[]) public delegations;
+
+// Checkpoints of voting power for an account.
+mapping(address => ERC20Votes.Checkpoint[]) internal _checkpoints;
+
+// Total supply checkpoints of the GovernanceToken.
+ERC20Votes.Checkpoint[] internal _totalSupplyCheckpoints;
+```
+
+## Constants
+
+The `GovernanceDelegation` contract MUST define the following constants:
+
+| Name                     | Value     | Description                                              |
+|--------------------------|-----------|----------------------------------------------------------|
+| `MAX_DELEGATIONS`        | `20`      | The maximum number of delegations allowed based on gas estimates of the worst-case scenario of an unoptimized version of the `_delegate` function |
+| `DENOMINATOR`            | `10_000`  | The denominator used for relative delegations.           |
+
+## Types
+
+The `GovernanceDelegation` contract MUST define the following types:
+
+### `Delegation`
+
+`Delegation` defines the parameters for voting power delegations, encapsulated in the following struct:
+
+```solidity
+struct Delegation {
+  AllowanceType allowanceType;
+  address delegatee;
+  uint256 amount;
+}
+```
+
+| Name                     | Type            | Description                                                             |
+|--------------------------|-----------------|-------------------------------------------------------------------------|
+| `allowanceType`          | `AllowanceType` | Type of allowance (e.g., absolute or relative).                         |
+| `delegatee`              | `address`       | The address of the delegatee receiving the voting power.                |
+| `amount`                 | `uint256`       | Amount of votes delegated, denomination depending on `allowanceType`.   |
+
+### `AllowanceType`
+
+Delegations can either be absolute or relative, represented by the following:
+
+```solidity
+enum AllowanceType {
+  Absolute,
+  Relative
+}
+```
+
+| Name        | Number | Description                                                                                 |
+|-------------|--------|---------------------------------------------------------------------------------------------|
+| `Absolute`  | `0`    | The amount of votes delegated is fixed and denominated in the `GovernanceToken`'s decimals  |
+| `Relative`  | `1`    | The amount of votes delegated is relative and denominated in percentages.                   |
+
+### `DelegationAdjustment`
+
+`DelegationAdjustment` act as temporary storage when adjusting the delegation state of an account. This struct is defined
```

Comment:
If this is an internal type it's best not to define it. We only need to define things that are part of the public interface.